### PR TITLE
[TT-1942] mw_jwt: implement method to extract OAuth client ID from JWT tokens

### PR DESCRIPTION
**Pull request by matiasinsaurralde**
## Description

This PR modifies the JWT middleware to be able to extract OAuth Client ID from JWT tokens.
Custom claims are used by identity providers, this includes common claim names for IDPs that are fully tested in our DCR flow.

Once the OAuth Client ID is extracted it's added to the session object, this ensures that a proper analytics record is generated.

In the future we could take two approaches: (i) keep extending the list for other common claim names or (ii) make this configurable.

## How This Has Been Tested
- Set a DCR enabled API.
- Register a client through the developer portal.
- Generate a token using the IDP token endpoint.
- Send a request to Tyk.
- Explore the log browser, the analytics record should contain the OAuth Client ID.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [x] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

File: gateway/mw_jwt.go [hunk 1]
```diff
@@ -547,6 +568,7 @@ func (k *JWTMiddleware) processCentralisedJWT(r *http.Request, token *jwt.Token)
 		}
 	}
 
+	session.OauthClientID = oauthClientID
```
Comment (buger) [inline gateway/mw_jwt.go:571]: There is really easy final step we can do:
```
if client, err := k.Spec.OAuthManager.OsinServer.Storage.GetClient(session.OauthClientID); err == nil {
  if developer_id, found := client.MetaData['developer_id']; found {
    session.DeveloperId = developer_id
  }
}
```

**Review by sredxny — APPROVED** [review]
So far LGTM

**Review by tbuchaillot — COMMENTED** [review]
+1 to @buger comments but overall LGTM. If we can add some unit tests to this it would be great.

File: gateway/mw_jwt.go (hunk 1 above)
Comment (matiasinsaurralde) [reply gateway/mw_jwt.go:571]: I've been able to get this working and I'm guessing the goal is to make it available inside the session metadata object.
I don't think this change would benefit the analytics side for now ([PR](https://github.com/TykTechnologies/tyk-analytics/pull/2245)) as we don't use this value on the analytics record itself, only OAuth client ID is used.

Comment (matiasinsaurralde) [reply gateway/mw_jwt.go:573]: Solved

**Comment by matiasinsaurralde** [comment]
Added a commit that properly handles the developer ID injection into session metadata in order to enable the logic that's implemented here:
https://github.com/TykTechnologies/tyk/blob/master/gateway/handler_success.go#L122

As expected the analytics record contains the developer ID:
Captura de Pantalla 2021-04-08 a la(s) 02 46 02

**Pull request merged**